File: leetcode/0238-product-of-array-except-self/python/product-of-array-except-self.py
from math import prod

def a_productExceptSelf(nums):
    answer = []
    for i in range(len(nums)):
        product = 1
        for j in range(len(nums)):
            if j != i:
                product *= nums[j]
        answer.append(product)
    return answer

# same as before
def b_productExceptSelf(nums):
    answer = []
    for i in range(len(nums)):
        product = prod(nums[:i]) * prod(nums[i + 1:])
        answer.append(product)
    return answer

def c_productExceptSelf(nums):
    product = 1
    answer = []
    for n in nums:
        product *= n
    for n in nums:
        answer.append(int(product / n))
    return answer

def d_productExceptSelf(nums):
    prefix = [1] * len(nums)
    suffix = [1] * len(nums)

    for i in range(1, len(nums)):
        prefix[i] = prefix[i - 1] * nums[i - 1]

    for i in range(len(nums) - 2, -1, -1):
        suffix[i] = suffix[i + 1] * nums[i + 1]
    
    return [prefix[i] * suffix[i] for i in range(len(nums))]

def e_productExceptSelf(nums):
    answer = [1] * len(nums)
    prefix = 1
    for i in range(len(nums)):
        answer[i] *= prefix
        prefix *= nums[i]
    suffix = 1
    for i in range(len(nums) - 1, -1, -1):
        answer[i] *= suffix
        suffix *= nums[i]
    return answer

print(e_productExceptSelf([1, 2, 3, 4]))

File: leetcode/0238-product-of-array-except-self/cpp/product-of-array-except-self.cpp
#include <iostream>
#include <vector>

using namespace std;

// brute-force
vector<int> productExceptSelf(vector<int>& nums) {
    vector<int> answer;
    for (int i = 0; i < nums.size(); i++) {
        int product = 1;
        for (int j = 0; j < nums.size(); j++) {
            if (j != i) {
                product *= nums[j];
            }
        }
        answer.push_back(product);
    }
    return answer;
}

// not in place solution
vector<int> productExceptSelf(const vector<int>& nums) {
    int n = nums.size();
    vector<int> prefix(n, 1);
    vector<int> suffix(n, 1);
    for (int i = 1; i < n; ++i) {
        prefix[i] = prefix[i - 1] * nums[i - 1];
    }
    for (int i = n - 2; i >= 0; --i) {
        suffix[i] = suffix[i + 1] * nums[i + 1];
    }
    vector<int> result(n);
    for (int i = 0; i < n; ++i) {
        result[i] = prefix[i] * suffix[i];
    }
    return result;
}

// optimal solution
vector<int> productExceptSelf(vector<int>& nums) {
    vector<int> answer(nums.size(), 1);
    int prefix = 1;
    for (int i = 0; i < nums.size(); i++) {
        answer[i] *= prefix;
        prefix *= nums[i];
    }
    int suffix = 1;
    for (int i = nums.size() - 1; i >= 0; i--) {
        answer[i] *= suffix;
        suffix *= nums[i];
    }
    return answer;
}


   vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> output(n);
        output[0] = 1;
        for(int i=1; i<n; i++){
            output[i] = output[i-1] * nums[i-1];
        }
        int right = 1;
        for(int i=n-1; i>=0; i--){
            output[i] *= right;
            right *= nums[i];
        }
        return output;
    }

